crates/base-db/src/editioned_file_id.rs

```diff
@@ -222,7 +222,18 @@ const _: () = {
                         .relevant_crates(editioned_file_id.file_id())
                         .first()
                         .copied()
-                        .unwrap_or_else(|| db.all_crates()[0]);
+                        .or_else(|| db.all_crates().first().copied())
+                        .unwrap_or_else(|| {
+                            // What we're doing here is a bit fishy. We rely on the fact that we only need
+                            // the crate in the item tree, and we should not create an `EditionedFileId`
+                            // without a crate except in cases where it does not matter. The chances that
+                            // `all_crates()` will be empty are also very slim, but it can occur during startup.
+                            // In the very unlikely case that there is a bug and we'll use this crate, Salsa
+                            // will panic.
+
+                            // SAFETY: 0 is less than `Id::MAX_U32`.
+                            salsa::plumbing::FromId::from_id(unsafe { salsa::Id::from_index(0) })
+                        });
```

Comment:
Do we really need that previous change (making the function non-optional). Not too fond of this hack, this might populate queries that aren't necessary if this value makes it through. I am also not too happy with the sheer amount of hacks we are slowly accumulating.

Reply:
That commit didn't introduce any hack, it just unified the majority of callers that did not want an optional `EditionedFileId`. The hack (that you also pointed at in the review) is in b8b33caaba109246a7eb1fffc744d78cb9aa5367, and I'm also not too fond of it, but I believe it's a necessary evil, as we really don't have anything to do if `all_crates()` is empty. I can change the fake id to `Id::MAX_U32 - 1`, though, which will pretty much guarantee a panic in case of a misuse and not populate some caches.